### PR TITLE
Julianschill better acceleration control

### PR DESCRIPTION
**Pull request by julianschill**

### CODE REVIEW

File: src/libslic3r/ExtrusionEntity.hpp [hunk 1]
```diff
@@ -120,6 +126,11 @@ inline bool is_solid_infill(ExtrusionRole role)
         || role == erIroning;
 }
 
+inline bool is_top_solid_infill(ExtrusionRole role)
+{
+    return role == erTopSolidInfill;
+}
```
Comment (supermerill) [inline src/libslic3r/ExtrusionEntity.hpp:132]: I think you can add erIroning here.
Also, why add it if you don't use it?

File: src/libslic3r/PrintConfig.cpp [hunk 2]
```diff
@@ -4414,15 +4465,22 @@ void PrintConfigDef::init_fff_params()
...
     def->set_default_value(new ConfigOptionFloatOrPercent(1500, false));
 
+    def = this->add("travel_deceleration_use_target", coBool);
+    def->label = L("Decelerate with target acceleration");
+    def->full_label = L("Use target acceleration for travel deceleration");
+    def->category = OptionCategory::speed;
+    def->tooltip = L("If selected, the deceleration of a travel will use the acceleration value of the extrusion that will be printed after it (if any) ");
+    def->mode = comExpert;
+    def->set_default_value(new ConfigOptionBool(true));
+
```
Comment (supermerill) [inline src/libslic3r/PrintConfig.cpp:4483]: Is there a case where we don't want it?

File: src/libslic3r/ExtrusionEntity.hpp [hunk 3]
```diff
@@ -95,12 +95,18 @@ enum ExtrusionLoopRole : uint16_t {
...
         || role == erThinWall
         || role == erOverhangPerimeter;
 }
 
+inline bool is_external_perimeter(ExtrusionRole role)
+{
+    return role == erExternalPerimeter
+        || role == erThinWall;
+}
+
```
Comment (supermerill) [inline src/libslic3r/ExtrusionEntity.hpp:109]: same as for is_top_infill, don't add things you don't use.

File: src/libslic3r/GCode.cpp [hunk 4]
```diff
@@ -3938,63 +3938,115 @@ std::string GCode::_before_extrude(const ExtrusionPath &path, const std::string
...
+
+        switch (path.role()){
+            case erPerimeter:
+                if (m_config.perimeter_acceleration.value >= 0)
+                    acceleration = m_config.perimeter_acceleration.get_abs_value(acceleration);
+                break;
+            case erExternalPerimeter:
+                if (m_config.external_perimeter_acceleration.value >= 0)
+                    acceleration = m_config.external_perimeter_acceleration.get_abs_value(acceleration);
+                break;
```
Comment (supermerill) [inline src/libslic3r/GCode.cpp:3953]: no thin walls, so?

File: src/libslic3r/GCode.cpp [hunk 5]
```diff
@@ -3938,63 +3938,115 @@ std::string GCode::_before_extrude(const ExtrusionPath &path, const std::string
...
+            case erCustom:
+            case erMixed:
+            case erCount:
+            case erNone:
+            default:
+                break;
         }
+
+        if (this->on_first_layer() && m_config.first_layer_acceleration.value > 0) {
+            acceleration = m_config.first_layer_acceleration.get_abs_value(acceleration);
```
Comment (supermerill) [inline src/libslic3r/GCode.cpp:3992]: maybe it's better to use a min afterwards? If you have a 800 accel for the external perimeter, 2000 for perimeter and set 1200 for the first layer, you may don't want to increase your external perimeter accel. I don't think people want to increase it for the first layer?

**Review by supermerill — COMMENTED** [review]
Good job overall.

**Comment by julianschill** [comment]
Sorry, I created the PR wrongly. Thanks for the comments, though. It is not done yet.

File: src/libslic3r/ExtrusionEntity.hpp (hunk 1 above)
Comment (julianschill) [reply src/libslic3r/ExtrusionEntity.hpp:132]: I will remove those functions again, they are not used anymore

File: src/libslic3r/PrintConfig.cpp (hunk 2 above)
Comment (julianschill) [reply src/libslic3r/PrintConfig.cpp:4402]: I have a request from a User who is just interested in the accelerations while extruding, not so much about ringing.

File: src/libslic3r/ExtrusionEntity.hpp (hunk 3 above)
Comment (julianschill) [reply src/libslic3r/ExtrusionEntity.hpp:109]: This was WIP. Will be removed

File: src/libslic3r/GCode.cpp (hunk 4 above)
Comment (julianschill) [reply src/libslic3r/GCode.cpp:3883]: Will consider thin walls

File: src/libslic3r/GCode.cpp (hunk 5 above)
Comment (julianschill) [reply src/libslic3r/GCode.cpp:3922]: First layer always gets printed with the first layer acceleration, regardless of the extrusion role. If you set a percentage you can slow it down and the percentage will be relative to the respective extrusion role

**Comment by julianschill** [comment]
I will consider your comments and create a new one, when its done.

File: src/libslic3r/GCode.cpp (hunk 5 above)
Comment (supermerill) [reply src/libslic3r/GCode.cpp:3922]: ok

**Pull request closed without merging**